language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,245 @@
+            label,
+            to,
+        };
+        self.0.get_mut(&from).unwrap().insert(new_edge);
+    }
+
+    /// adds a strong edge and factors other edges coming out of `from` with respect to the new edge
+    pub fn add_strong_edge(&mut self, from: Nonce, label: Label<T>, to: Nonce) {
+        checked_precondition!(self.0.contains_key(&from));
+        checked_precondition!(self.0.contains_key(&to));
+        checked_precondition!(self.0[&to].is_empty());
+        checked_precondition!(label.len() <= 1);
+
+        let new_edge = Edge {
+            edge_type: EdgeType::Strong,
+            label: label.clone(),
+            to,
+        };
+        self.0.remove(&to).unwrap();
+        if label.is_empty() {
+            let from_edge_set = self.0.remove(&from).unwrap();
+            let mut new_from_edge_set = BTreeSet::new();
+            new_from_edge_set.insert(new_edge);
+            self.0.insert(from, new_from_edge_set);
+            self.0.insert(to, from_edge_set);
+        } else {
+            let mut edge_set: Vec<Edge<T>> = self.0[&from]
+                .iter()
+                .filter(|x| !x.label.is_empty() && x.label[0] == label[0])
+                .cloned()
+                .collect();
+            let from_edge_set = self.0.get_mut(&from).unwrap();
+            edge_set.iter().for_each(|x| {
+                from_edge_set.remove(x);
+            });
+            from_edge_set.insert(new_edge);
+            edge_set.iter_mut().for_each(|x| {
+                x.label.remove(0);
+            });
+            self.0.insert(to, edge_set.into_iter().collect());
+        }
+    }
+
+    /// removes `nonce` and appropriately concatenates each incoming edge with each outgoing edge of `nonce`
+    pub fn remove_nonce(&mut self, nonce: Nonce) {
+        checked_assume!(self.invariant());
+        let nonce_edge_set = self.0.remove(&nonce).unwrap();
+        let mut removed_edges = BTreeMap::new();
+        for (n, es) in &self.0 {
+            removed_edges.insert(
+                n.clone(),
+                es.iter()
+                    .filter(|x| x.to == nonce)
+                    .cloned()
+                    .collect::<Vec<_>>(),
+            );
+        }
+        for (n, es) in &removed_edges {
+            es.iter().for_each(|removed_edge| {
+                let n_edge_set_ref = self.0.get_mut(n).unwrap();
+                n_edge_set_ref.remove(removed_edge);
+                nonce_edge_set.iter().for_each(|nonce_edge| {
+                    if removed_edge.edge_type == EdgeType::Strong {
+                        let mut new_label = vec![];
+                        new_label.append(&mut removed_edge.label.clone());
+                        new_label.append(&mut nonce_edge.label.clone());
+                        let edge = Edge {
+                            edge_type: nonce_edge.edge_type,
+                            label: new_label,
+                            to: nonce_edge.to,
```

Comment:
Could [struct update syntax](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) be useful here?

Reply:
I prefer the explicit construction here. The construction of these edges is very delicate

Reply:
Let us stick with the explicit syntax for now.